### PR TITLE
OPEN: FP integration

### PR DESCRIPTION
**Pull request by Dequino**
- Reintegrated immediate float in AbstractDataTypes​
- Reintegrated float immediate in DeeployTypes
- Reintegrated bfloat16, float32, FloatDataTypes to DataTypes
- Reintegrated testImmediatePromotionFloat
- Reintegrated FloatAdder, tested FP32 on generic platform
- Also tested on siracusa, output buffer is the correct but gvsoc also returns an error​  

Future work:
- Integrate PULP-Trainlib floating point kernels as third-party library

### CODE REVIEW

**Comment by Victor-Jung** [comment]
Hi Alberto! Since this PR hasn't been merged in the GitLab repo, it's not really a reintegration. Could you make a more detailed description of the proposed feature? 

Additionally, I think it's best not to add the PULP-TrainLib kernel integration into this PR.

**Comment by Dequino** [comment]
Hi Victor,

Sure, we can move the pulp-trainlib integration in a separate PR.

I'll be more through on the contents and added features:

Feature addition: Float Immediate abstract type
I've added the Float Immediate to the abstract types to be integrated in deeploy. 
At the moment, the workframe has been developed focusing on deploying networks quantized with integer data. 
It may be useful to future proof the framework by enabling integration of arbitrary floating-point precision formats. To do so, the first step was to define FloatImmediate class in the AbstractDataTypes, more precisely I've defined a method that checks if an immediate value can be represented by a float format with an arbitrary number of bits reserved for the exponent and fraction part.

To test the validity of this method, I've integrated the testImmediatePromotionFloat to the testTypes.py script, in which we test different immediate values to be represented in float32 and bfloat16 formats, which have been added as FloatImmediate extensions in Deeploy/CommonExtensions/DataTypes. If we want to add a new floating point precision format, we can simply define it here by specifying its total width, number of reserved bits for the fraction, and number of reserved bits for the exponent.

After that, I've tested the correct integration of floats by testing a simple deployment of a dummy model doing an addition between two float tensors. The test model and data is defined in Test/FloatAdder. I've had to edit a number of files to enable float data types, including:

- Deeploy/DeeployTypes
- Generic/Bindings
- Generic/TypeCheckers
- Deeploytest/GenerateNetowrk
- testUtils/typeMapping

I've also added some minor edits to the documentation, as some line commands were incorrect or incomplete.

If you have more in depth questions about the changes, let me know

**Review by Scheremo — COMMENTED** [review]
Hi Alberto! I had a look at your proposed changes, and I think they are overall in okay shape. My main advise would be to rely on standard library functions for mantissa and exponent extraction (they have python bindings), as the current implementation is hard to follow and seems to be prone to edge cases.

File: Deeploy/AbstractDataTypes.py [hunk 1]
```diff
@@ -234,6 +235,139 @@ def checkValue(cls, value: Union[int, Iterable[int]], ctxt: Optional[_NetworkCon
         return True
 
 
+class FloatImmediate(Immediate[Union[float, Iterable[float]], _ImmediateType]):
+    typeFraction: int  #: int: Represents the number of bits reserved for the fraction part
+    typeExponent: int  #: int: Represents the number of bits reserved for the exponent part
+    signed: bool  #: bool: Represents whether the underlying float is signed or unsigned (should be removed)
```
[the remark this answers is not in the document]
Comment (Dequino) [reply Deeploy/AbstractDataTypes.py:241]: Sure, float numbers are all signed after all, I just put the bool here to make sure it wouldn't conflict with anything else in the framework

**Comment by FrancescoConti** [comment]
Closing as superseded by (and included into) #12
